Extreme memory usage with large blocks of text

Is it normal/expected to see this kind of memory usage? Memory is shown at the top of the screencaps. 

The first one is a modified version of one of the demos which adds 4,000 lines. Memory usage on that one spikes to about twice what's shown here, and then drops down. It also functions fine. 

![rtfx1](https://user-images.githubusercontent.com/11338660/31857412-0cb00376-b6ac-11e7-80e8-bb4122d7095e.png)

This second screenshot is with 400,000 extra rows of text added. Demo becomes unusable with this much text. When I add a single letter to a row, it freezes for a few seconds before showing the character. Scrolling still works okay, but the rendering speed isn't great. Memory usage... 2GB? :exclamation: :exclamation: :exclamation: It also fluctuates wildly (up to 1GB or so) when scrolling, which is only going to make things worse.

![rtfx2](https://user-images.githubusercontent.com/11338660/31857413-0f37d6a0-b6ac-11e7-8bd6-cdd5a1305a29.png)

I made every line different on purpose since it looks like some optimizations are going on under the hood, but even without that I still see this behavior. 

I'm running on Linux. Oracle JRE 8u144

It seems a bit better doing the above, at least memory wise. Side effect I see here is that highlighting doesn't work during scrolling and takes a second to engage once scrolling stops. Not ideal but maybe okay. Still kinda sluggish when paging up/down. Another thing I noticed was this Error being thrown (so, maybe I have something screwed up), although it seems to run fine regardless. 

> Exception in thread "JavaFX Application Thread" java.lang.AssertionError: Unreachable code
> 	at org.fxmisc.richtext.GenericStyledArea.visibleParToAllParIndex(GenericStyledArea.java:786)
> 	at org.fxmisc.richtext.demo.ScalabilityTest.lambda$start$2(ScalabilityTest.java:145)
> 	at org.reactfx.util.NonAccumulativeStreamNotifications.lambda$head$0(NotificationAccumulator.java:134)
> 	at org.reactfx.ObservableBase.notifyObservers(ObservableBase.java:68)
> 	at org.reactfx.ObservableBase.notifyObservers(ObservableBase.java:57)
> 	at org.reactfx.ProperEventStream.emit(ProperEventStream.java:18)
> 	at org.reactfx.util.NonAccumulativeStreamNotifications.lambda$head$0(NotificationAccumulator.java:134)
> 	at org.reactfx.ObservableBase.notifyObservers(ObservableBase.java:68)
> 	at org.reactfx.ObservableBase.notifyObservers(ObservableBase.java:57)
> 	at org.reactfx.ProperEventStream.emit(ProperEventStream.java:18)
> 	at org.reactfx.SuccessionReducingStream.handleTimeout(SuccessionReducingStream.java:73)
> 	at org.reactfx.util.FxTimer.lambda$restart$0(FxTimer.java:89)
> 	at com.sun.scenario.animation.shared.TimelineClipCore.visitKeyFrame(TimelineClipCore.java:239)
> 	at com.sun.scenario.animation.shared.TimelineClipCore.playTo(TimelineClipCore.java:180)
> 	at javafx.animation.Timeline.impl_playTo(Timeline.java:176)
> ...

ScalabilityTest.java:145: 
int firstVisibleParIdx = codeArea.visibleParToAllParIndex(0);
That only showed up twice right at startup, so maybe just a symptom of how that demo initializes. 

I did remove this bit from the original example above: 
```
//        codeArea.richChanges()
//                .filter(ch -> !ch.getInserted().equals(ch.getRemoved())) // XXX
//                .subscribe(change -> {
//                    codeArea.setStyleSpans(0, computeHighlighting(getVisibleText(codeArea)));
//                });
```

I have to admit though, the functional reactive development style is painful for me. Either I'm getting old and stubborn or it's just unfamiliarity, but it really obfuscates tracking what the code is _doing_, for me. Between that and the fact that I really don't _need_ highlighting (or even editing), maybe I'll just skip it.

--- a/richtextfx/src/main/java/org/fxmisc/richtext/ParagraphText.java
+++ b/richtextfx/src/main/java/org/fxmisc/richtext/ParagraphText.java
@@ -1,5 +1,6 @@
 package org.fxmisc.richtext;
 
+import java.lang.ref.WeakReference;
 import java.util.Arrays;
 import java.util.Collection;
 import java.util.HashMap;
@@ -18,6 +19,7 @@ import java.util.function.UnaryOperator;
 import javafx.beans.property.ObjectProperty;
 import javafx.beans.property.SimpleObjectProperty;
 import javafx.beans.value.ChangeListener;
+import javafx.beans.value.ObservableValue;
 import javafx.collections.FXCollections;
 import javafx.collections.MapChangeListener;
 import javafx.collections.ObservableMap;
@@ -93,48 +95,11 @@ class ParagraphText<PS, SEG, S> extends TextFlowExt {
         Val<Double> leftInset = Val.map(insetsProperty(), Insets::getLeft);
         Val<Double> topInset = Val.map(insetsProperty(), Insets::getTop);
 
-        ChangeListener<IndexRange> requestLayout1 = (obs, ov, nv) -> requestLayout();
+        ChangeListener<IndexRange> requestLayout1 = new SelectionRangeChangeListener<>(this);
+        selections.addListener(new SelectionsSetListener<>(leftInset, requestLayout1, topInset, this));
 
-        selections.addListener((MapChangeListener.Change<? extends Selection<PS, SEG, S>, ? extends SelectionPath> change) -> {
-            if (change.wasAdded()) {
-                SelectionPath p = change.getValueAdded();
-                p.rangeProperty().addListener(requestLayout1);
-
-                p.layoutXProperty().bind(leftInset);
-                p.layoutYProperty().bind(topInset);
-
-                getChildren().add(selectionShapeStartIndex, p);
-                updateSingleSelection(p);
-            } else if (change.wasRemoved()) {
-                SelectionPath p = change.getValueRemoved();
-                p.rangeProperty().removeListener(requestLayout1);
-
-                p.layoutXProperty().unbind();
-                p.layoutYProperty().unbind();
-
-                getChildren().remove(p);
-            }
-        });
-
-        ChangeListener<Integer> requestLayout2 = (obs, ov, nv) -> requestLayout();
-        carets.addListener((SetChangeListener.Change<? extends CaretNode> change) -> {
-            if (change.wasAdded()) {
-                CaretNode caret = change.getElementAdded();
-                caret.columnPositionProperty().addListener(requestLayout2);
-                caret.layoutXProperty().bind(leftInset);
-                caret.layoutYProperty().bind(topInset);
-
-                getChildren().add(caret);
-                updateSingleCaret(caret);
-            } else if (change.wasRemoved()) {
-                CaretNode caret = change.getElementRemoved();
-                caret.columnPositionProperty().removeListener(requestLayout2);
-                caret.layoutXProperty().unbind();
-                caret.layoutYProperty().unbind();
-
-                getChildren().remove(caret);
-            }
-        });
+        ChangeListener<Integer> requestLayout2 = new CaretPositionChangeListener<>(this);
+        carets.addListener(new CaretsChangeListener<>(leftInset, requestLayout2, topInset, this));
 
         // XXX: see the note at highlightTextFill
 //        highlightTextFill.addListener(new ChangeListener<Paint>() {
@@ -435,6 +400,131 @@ class ParagraphText<PS, SEG, S> extends TextFlowExt {
         updateAllCaretShapes();
         updateAllSelectionShapes();
         updateBackgroundShapes();
+    }
+
+    private static final class SelectionsSetListener<PS, SEG, S> implements
+                MapChangeListener<Selection<PS, SEG, S>, SelectionPath> {
+        private final Val<Double> leftInset;
+        private final ChangeListener<IndexRange> requestLayout1;
+        private final Val<Double> topInset;
+        private final WeakReference<ParagraphText<PS, SEG, S>> ref;
+
+        public SelectionsSetListener(
+                    Val<Double> leftInset,
+                    ChangeListener<IndexRange> requestLayout1,
+                    Val<Double> topInset,
+                    ParagraphText<PS, SEG, S> paragraphText) {
+            this.leftInset = leftInset;
+            this.requestLayout1 = requestLayout1;
+            this.topInset = topInset;
+            ref = new WeakReference<>(paragraphText);
+        }
+
+        @Override
+        public void onChanged(
+                    javafx.collections.MapChangeListener.Change<? extends Selection<PS, SEG, S>, ? extends SelectionPath> change) {
+            ParagraphText<PS, SEG, S> paragraphText = ref.get();
+            if (null == paragraphText) {
+                change.getMap().removeListener(this);
+                return;
+            }
+
+            if (change.wasAdded()) {
+                SelectionPath p = change.getValueAdded();
+                p.rangeProperty().addListener(requestLayout1);
+
+                p.layoutXProperty().bind(leftInset);
+                p.layoutYProperty().bind(topInset);
+
+                paragraphText.getChildren().add(paragraphText.selectionShapeStartIndex, p);
+                paragraphText.updateSingleSelection(p);
+            } else if (change.wasRemoved()) {
+                SelectionPath p = change.getValueRemoved();
+                p.rangeProperty().removeListener(requestLayout1);
+
+                p.layoutXProperty().unbind();
+                p.layoutYProperty().unbind();
+
+                paragraphText.getChildren().remove(p);
+            }
+        }
+    }
+
+    private static final class SelectionRangeChangeListener<PS, SEG, S> implements ChangeListener<IndexRange> {
+        private final WeakReference<ParagraphText<PS, SEG, S>> ref;
+
+        public SelectionRangeChangeListener(ParagraphText<PS, SEG, S> paragraphText) {
+            ref = new WeakReference<>(paragraphText);
+        }
+
+        @Override
+        public void changed(ObservableValue<? extends IndexRange> observable, IndexRange oldValue, IndexRange newValue) {
+            if (null == ref.get()) {
+                observable.removeListener(this);
+            } else {
+                ref.get().requestLayout();
+            }
+        }
+    }
+
+    private static final class CaretPositionChangeListener<PS, SEG, S> implements ChangeListener<Integer> {
+        private final WeakReference<ParagraphText<PS, SEG, S>> ref;
+
+        public CaretPositionChangeListener(ParagraphText<PS, SEG, S> paragraphText) {
+            ref = new WeakReference<>(paragraphText);
+        }
+
+        @Override
+        public void changed(ObservableValue<? extends Integer> observable, Integer oldValue, Integer newValue) {
+            if (null == ref.get()) {
+                observable.removeListener(this);
+            } else {
+                ref.get().requestLayout();
+            }
+        }
+    }
+
+    private static final class CaretsChangeListener<PS, SEG, S> implements SetChangeListener<CaretNode> {
+        private final Val<Double> leftInset;
+        private final ChangeListener<Integer> requestLayout2;
+        private final Val<Double> topInset;
+        private final WeakReference<ParagraphText<PS, SEG, S>> ref;
+
+        private CaretsChangeListener(
+                    Val<Double> leftInset,
+                    ChangeListener<Integer> requestLayout2,
+                    Val<Double> topInset,
+                    ParagraphText<PS, SEG, S> paragraphText) {
+            ref = new WeakReference<>(paragraphText);
+            this.leftInset = leftInset;
+            this.requestLayout2 = requestLayout2;
+            this.topInset = topInset;
+        }
+
+        @Override
+        public void onChanged(Change<? extends CaretNode> change) {
+            ParagraphText<PS, SEG, S> paragraphText = ref.get();
+            if (null == paragraphText) {
+                change.getSet().removeListener(this);
+                return;
+            }
+            if (change.wasAdded()) {
+                CaretNode caret = change.getElementAdded();
+                caret.columnPositionProperty().addListener(requestLayout2);
+                caret.layoutXProperty().bind(leftInset);
+                caret.layoutYProperty().bind(topInset);
+
+                paragraphText.getChildren().add(caret);
+                paragraphText.updateSingleCaret(caret);
+            } else if (change.wasRemoved()) {
+                CaretNode caret = change.getElementRemoved();
+                caret.columnPositionProperty().removeListener(requestLayout2);
+                caret.layoutXProperty().unbind();
+                caret.layoutYProperty().unbind();
+
+                paragraphText.getChildren().remove(caret);
+            }
+        }
     }
 
     private static class CustomCssShapeHelper<T> {

